File: android/src/main/java/com/sahha/capacitor/SahhaPlugin.kt
package com.sahha.capacitor;

import ai.sahha.api.Sahha
import ai.sahha.api.biomarkers.SahhaBiomarkerCategory
import ai.sahha.api.biomarkers.SahhaBiomarkerType
import ai.sahha.api.demographic.SahhaDemographic
import ai.sahha.api.notifications.SahhaNotificationConfiguration
import ai.sahha.api.score.SahhaScoreType
import ai.sahha.api.sensors.SahhaSensor
import ai.sahha.api.settings.SahhaEnvironment
import ai.sahha.api.settings.SahhaFramework
import ai.sahha.api.settings.SahhaSettings
import android.content.Context
import android.util.Log
import androidx.activity.ComponentActivity
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.gson.GsonBuilder
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializer
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.ZonedDateTime

private const val TAG = "SahhaPlugin"

@CapacitorPlugin(name = "Sahha")
public class SahhaPlugin : Plugin() {

    @PluginMethod
    fun configure(call: PluginCall) {

        val settings: JSObject? = call.getObject("settings")

        if (settings == null) {
            call.reject("Sahha configure settings parameter is missing")
            return
        }

        val environment: String? = settings.getString("environment")
        if (environment == null) {
            call.reject("Sahha configure settings environment parameter is missing")
            return
        }

        Log.d(TAG, "Sahha.configure environment: $environment")
        val sahhaEnvironment: SahhaEnvironment
        try {
            val availableEnvironments = SahhaEnvironment.entries
            sahhaEnvironment = availableEnvironments.find { it.name.equals(environment, ignoreCase = true) }
                ?: SahhaEnvironment.valueOf(environment.uppercase())
            Log.d(TAG, "Sahha.configure resolved to: ${sahhaEnvironment.name}")
        } catch (e: Exception) {
            val available = SahhaEnvironment.entries.joinToString(", ") { it.name }
            Log.e(TAG, "Invalid environment: $environment. Available: $available", e)
            call.reject("Sahha configure settings environment parameter is not valid. Available: $available")
            return
        }

        // Notification config
        var sahhaNotificationConfiguration = SahhaNotificationConfiguration.DEFAULT
        try {
            settings.getJSObject("notificationSettings")?.also { nSettings ->
                val icon = nSettings.getString("icon")
                val title = nSettings.getString("title")
                val shortDescription = nSettings.getString("shortDescription")

                sahhaNotificationConfiguration = SahhaNotificationConfiguration(
                    stringToDrawableResource(
                        context,
                        icon
                    ) ?: 0,
                    title ?: "title",
                    shortDescription ?: "shortDescription",
                )
            }
        } catch (e: IllegalArgumentException) {
            call.reject("Sahha.configure() notification config is not valid")
            return
        }
        // Notification config ends

        val sahhaSettings = SahhaSettings(
            sahhaEnvironment,
            SahhaFramework.CAPACITOR,
            sahhaNotificationConfiguration
        )

        val activity = activity as? ComponentActivity
        if (activity == null) {
            call.reject("Sahha configure app is missing")
        } else {
            Sahha.configure(activity, sahhaSettings) { error, success ->
                if (error != null) {
                    call.reject(error)
                } else {
                    val data = JSObject()
                    data.put("success", success)
                    call.resolve(data)
                }
            }
        }
    }

    @PluginMethod
    fun isAuthenticated(call: PluginCall) {
        val data = JSObject()
        data.put("success", Sahha.isAuthenticated)
        call.resolve(data);
    }

    @PluginMethod
    fun authenticate(call: PluginCall) {

        var appId: String? = call.getString("appId");
        if (appId == null) {
            call.reject("Sahha authenticate appId parameter is missing")
            return
        }

        var appSecret: String? = call.getString("appSecret");
        if (appSecret == null) {
            call.reject("Sahha authenticate appSecret parameter is missing")
            return
        }

        var externalId: String? = call.getString("externalId");
        if (externalId == null) {
            call.reject("Sahha authenticate externalId parameter is missing")
            return
        }

        Log.d(TAG, "Sahha.authenticate starting for appId: $appId, externalId: $externalId")
        Sahha.authenticate(appId, appSecret, externalId)
        { error, success ->
            if (error != null) {
                call.reject(error);
            } else {
                val data = JSObject()
                data.put("success", true)
                call.resolve(data);
            }
        }
    }

    @PluginMethod
    fun deauthenticate(call: PluginCall) {
        Sahha.deauthenticate() { error, success ->
            if (error != null) {
                call.reject(error)
            } else {
                val data = JSObject()
                data.put("success", success)
                call.resolve(data)
            }
        }
    }

    @PluginMethod
    fun getProfileToken(call: PluginCall) {
        val profileToken = Sahha.profileToken
        if (profileToken != null) {
            val data = JSObject()
            data.put("profileToken", profileToken)
            call.resolve(data)
        } else {
            call.reject("No profile token found")
        }
    }

    @PluginMethod
    fun authenticateToken(call: PluginCall) {

        var profileToken: String? = call.getString("profileToken");
        if (profileToken == null) {
            call.reject("Sahha authenticate profileToken parameter is missing")
            return
        }

        var refreshToken: String? = call.getString("refreshToken");
        if (refreshToken == null) {
            call.reject("Sahha authenticate refreshToken parameter is missing")
            return
        }

        Sahha.authenticate(profileToken, refreshToken)
        { error, success ->
            if (error != null) {
                call.reject(error);
            } else {
                val data = JSObject()
                data.put("success", success)
                call.resolve(data);
            }
        }
    }

    @PluginMethod
    fun getDemographic(call: PluginCall) {

        Sahha.getDemographic() { error, demographic ->
            if (error != null) {
                call.reject(error)
            } else if (demographic != null) {

                // Need to handle possible null values
                val gson = GsonBuilder().serializeNulls().create()

                val demographicJson: String = gson.toJson(demographic)
                Log.d("Sahha", demographicJson)
                val data = JSObject()
                data.put("demographic", demographicJson)
                call.resolve(data)
            } else {
                call.reject("Sahha demographic not available")
            }
        }
    }

    @PluginMethod
    fun postDemographic(call: PluginCall) {

        var demographic: JSObject? = call.getObject("demographic")

        if (demographic == null) {
            call.reject("Sahha demographic parameter is missing")
            return
        }

        val age: Int? = demographic.optInt("age")
        val gender: String? = demographic.optString("gender")
        val birthDate: String? = demographic.optString("birthDate")

        var sahhaDemographic = SahhaDemographic(
            age = if (age == 0) null else age,
            gender = gender?.ifEmpty { null },
            birthDate = birthDate?.ifEmpty { null },
        )
        Log.d("Sahha", sahhaDemographic.toString())

        Sahha.postDemographic(sahhaDemographic) { error, success ->
            if (error != null) {
                call.reject(error)
            } else {
                val data = JSObject()
                data.put("success", success)
                call.resolve(data)
            }
        }
    }


private fun parseSensors(call: PluginCall, methodName: String): MutableSet<SahhaSensor>? {
    val sensors: JSArray? = call.getArray("sensors")

    if (sensors == null) {
        call.reject("Sahha $methodName sensors parameter is missing")
        return null
    }

    val sahhaSensors = mutableSetOf<SahhaSensor>()
    try {
        for (i in 0 until sensors.length()) {
            val sensor = sensors.getString(i)
            val sahhaSensor = SahhaSensor.valueOf(sensor.uppercase())
            sahhaSensors.add(sahhaSensor)
        }
    } catch (e: IllegalArgumentException) {
        call.reject("Sahha sensor parameter is not valid")
        return null
    }

    return sahhaSensors
}

private fun resolveSensorStatus(call: PluginCall, sahhaSensors: Set<SahhaSensor>) {
    Sahha.getSensorStatus(sahhaSensors) { error, sensorStatus ->
        if (error != null) {
            call.reject(error)
        } else {
            val data = JSObject()
            data.put("status", sensorStatus)
            call.resolve(data)
        }
    }
}


  @PluginMethod
fun getSensorStatus(call: PluginCall) {
    val sahhaSensors = parseSensors(call, "getSensorStatus") ?: return
    resolveSensorStatus(call, sahhaSensors)
}

@PluginMethod
fun enableSensors(call: PluginCall) {
    val sahhaSensors = parseSensors(call, "enableSensors") ?: return

    Sahha.enableSensors(sahhaSensors) { error, _ ->
        if (error != null) {
            call.reject(error)
        } else {
            resolveSensorStatus(call, sahhaSensors)
        }
    }
}

    @PluginMethod
    fun getScores(call: PluginCall) {
        val types: JSArray? = call.getArray("types")
        val startDateEpochMilli: Long? = call.getLong("startDateTime")
        println("startDateTime provided: $startDateEpochMilli")
        val endDateEpochMilli: Long? = call.getLong("endDateTime")
        println("endDateTime provided: $endDateEpochMilli")

        if (types == null) {
            call.reject("Sahha getScores types parameter is missing")
            return
        }

        if (startDateEpochMilli == null) {
            call.reject("Sahha getScores startDateTime parameter is missing")
            return
        }

        if (endDateEpochMilli == null) {
            call.reject("Sahha getScores endDateTime parameter is missing")
            return
        }

        val sahhaScoreTypes: MutableSet<SahhaScoreType> = mutableSetOf<SahhaScoreType>()
        try {
            for (i in 0 until types.length()) {
                val type: String = types.getString(i)
                val scoreType = SahhaScoreType.valueOf(type.uppercase())
                sahhaScoreTypes.add(scoreType)
            }
        } catch (e: IllegalArgumentException) {
            call.reject("Sahha score type parameter is not valid")
            return
        }

        val startDateIsNotNull = startDateEpochMilli != null
        val endDateIsNotNull = endDateEpochMilli != null

        if (startDateIsNotNull && endDateIsNotNull) {
            val defaultZoneId = ZoneId.systemDefault()
            val startInstant = Instant.ofEpochMilli(startDateEpochMilli!!)
            val startLocalDateTime = LocalDateTime.ofInstant(startInstant, defaultZoneId)
            val endInstant = Instant.ofEpochMilli(endDateEpochMilli!!)
            val endLocalDateTime = LocalDateTime.ofInstant(endInstant, defaultZoneId)

            Sahha.getScores(
                sahhaScoreTypes,
                Pair(startLocalDateTime, endLocalDateTime)
            ) { error, value ->
                if (error != null) {
                    call.reject(error)
                } else {
                    val gson = GsonBuilder()
                        .registerTypeAdapter(ZonedDateTime::class.java,
                            JsonSerializer<ZonedDateTime> { src, _, _ ->
                                JsonPrimitive(src.toString())
                            }).create()
                    val valueJson = gson.toJson(value)
                    
                    val data = JSObject()
                    data.put("value", valueJson)
                    call.resolve(data)
                }
            }
        } else {
            Sahha.getScores(sahhaScoreTypes) { error, value ->
                if (error != null) {
                    call.reject(error)
                } else {
                    val gson = GsonBuilder()
                        .registerTypeAdapter(ZonedDateTime::class.java,
                            JsonSerializer<ZonedDateTime> { src, _, _ ->
                                JsonPrimitive(src.toString())
                            }).create()
                    val valueJson = gson.toJson(value)
                    
                    val data = JSObject()
                    data.put("value", valueJson)
                    call.resolve(data)
                }
            }
        }
    }

    @PluginMethod
    fun getBiomarkers(call: PluginCall) {
        val categories: JSArray? = call.getArray("categories")
        val types: JSArray? = call.getArray("types")
        val startDateEpochMilli: Long? = call.getLong("startDateTime")
        println("startDateTime provided: $startDateEpochMilli")
        val endDateEpochMilli: Long? = call.getLong("endDateTime")
        println("endDateTime provided: $endDateEpochMilli")

        if (categories == null) {
            call.reject("Sahha getBiomarkers categories parameter is missing")
            return
        }

        if (types == null) {
            call.reject("Sahha getBiomarkers types parameter is missing")
            return
        }

        if (startDateEpochMilli == null) {
            call.reject("Sahha getBiomarkers startDateTime parameter is missing")
            return
        }

        if (endDateEpochMilli == null) {
            call.reject("Sahha getBiomarkers endDateTime parameter is missing")
            return
        }

        val sahhaBiomarkerCategories: MutableSet<SahhaBiomarkerCategory> =
            mutableSetOf<SahhaBiomarkerCategory>()
        try {
            for (i in 0 until categories.length()) {
                val category: String = categories.getString(i)
                val biomarkerCategory = SahhaBiomarkerCategory.valueOf(category.uppercase())
                sahhaBiomarkerCategories.add(biomarkerCategory)
            }
        } catch (e: IllegalArgumentException) {
            call.reject("Sahha biomarker category parameter is not valid")
            return
        }

        val sahhaBiomarkerTypes: MutableSet<SahhaBiomarkerType> = mutableSetOf<SahhaBiomarkerType>()
        try {
            for (i in 0 until types.length()) {
                val type: String = types.getString(i)
                val biomarkerType = SahhaBiomarkerType.valueOf(type.uppercase())
                sahhaBiomarkerTypes.add(biomarkerType)
            }
        } catch (e: IllegalArgumentException) {
            call.reject("Sahha biomarker type parameter is not valid")
            return
        }

        val startDateIsNotNull = startDateEpochMilli != null
        val endDateIsNotNull = endDateEpochMilli != null

        if (startDateIsNotNull && endDateIsNotNull) {
            val defaultZoneId = ZoneId.systemDefault()
            val startInstant = Instant.ofEpochMilli(startDateEpochMilli!!)
            val startLocalDateTime = LocalDateTime.ofInstant(startInstant, defaultZoneId)
            val endInstant = Instant.ofEpochMilli(endDateEpochMilli!!)
            val endLocalDateTime = LocalDateTime.ofInstant(endInstant, defaultZoneId)

            Sahha.getBiomarkers(
                sahhaBiomarkerCategories,
                sahhaBiomarkerTypes,
                Pair(startLocalDateTime, endLocalDateTime)
            ) { error, value ->
                if (error != null) {
                    call.reject(error)
                } else {
                    val gson = GsonBuilder()
                        .registerTypeAdapter(ZonedDateTime::class.java,
                            JsonSerializer<ZonedDateTime> { src, _, _ ->
                                JsonPrimitive(src.toString())
                            }).create()
                    val valueJson = gson.toJson(value)
                    
                    val data = JSObject()
                    data.put("value", valueJson)
                    call.resolve(data)
                }
            }
        } else {
            Sahha.getBiomarkers(sahhaBiomarkerCategories, sahhaBiomarkerTypes) { error, value ->
                if (error != null) {
                    call.reject(error)
                } else {
                    val gson = GsonBuilder()
                        .registerTypeAdapter(ZonedDateTime::class.java,
                            JsonSerializer<ZonedDateTime> { src, _, _ ->
                                JsonPrimitive(src.toString())
                            }).create()
                    val valueJson = gson.toJson(value)
                    
                    val data = JSObject()
                    data.put("value", valueJson)
                    call.resolve(data)
                }
            }
        }
    }

    @PluginMethod
    fun getStats(call: PluginCall) {
        val sensor: String? = call.getString("sensor")
        println("sensor provided: $sensor")
        val startDateEpochMilli: Long? = call.getLong("startDateTime")
        println("startDateTime provided: $startDateEpochMilli")
        val endDateEpochMilli: Long? = call.getLong("endDateTime")
        println("endDateTime provided: $endDateEpochMilli")

        if (sensor == null) {
            call.reject("Sahha getStats sensor parameter is missing")
            return
        }

        if (startDateEpochMilli == null) {
            call.reject("Sahha getStats startDateTime parameter is missing")
            return
        }

        if (endDateEpochMilli == null) {
            call.reject("Sahha getStats endDateTime parameter is missing")
            return
        }

        val defaultZoneId = ZoneId.systemDefault()
        val startInstant = Instant.ofEpochMilli(startDateEpochMilli)
        val startLocalDateTime = LocalDateTime.ofInstant(startInstant, defaultZoneId)
        val endInstant = Instant.ofEpochMilli(endDateEpochMilli)
        val endLocalDateTime = LocalDateTime.ofInstant(endInstant, defaultZoneId)

        Sahha.getStats(
            sensor = SahhaSensor.valueOf(sensor.uppercase()),
            Pair(startLocalDateTime, endLocalDateTime)
        ) { error, value ->
            if (error != null) {
                if (error.contains("found", ignoreCase = true)) {
                    val data = JSObject()
                    data.put("value", "[]")
                    call.resolve(data)
                } else {
                    call.reject(error)
                }
            } else {
                val gson = GsonBuilder()
                    .registerTypeAdapter(ZonedDateTime::class.java,
                        JsonSerializer<ZonedDateTime> { src, _, _ ->
                            JsonPrimitive(src.toString())
                        }).create()

                val statsJson = gson.toJson(value)
                val data = JSObject()
                data.put("value", statsJson)
                call.resolve(data)
            }
        }
    }

    @PluginMethod
    fun getSamples(call: PluginCall) {
        val sensor: String? = call.getString("sensor")
        println("sensor provided: $sensor")
        val startDateEpochMilli: Long? = call.getLong("startDateTime")
        println("startDateTime provided: $startDateEpochMilli")
        val endDateEpochMilli: Long? = call.getLong("endDateTime")
        println("endDateTime provided: $endDateEpochMilli")

        if (sensor == null) {
            call.reject("Sahha getSamples sensor parameter is missing")
            return
        }

        if (startDateEpochMilli == null) {
            call.reject("Sahha getSamples startDateTime parameter is missing")
            return
        }

        if (endDateEpochMilli == null) {
            call.reject("Sahha getSamples endDateTime parameter is missing")
            return
        }

        val defaultZoneId = ZoneId.systemDefault()
        val startInstant = Instant.ofEpochMilli(startDateEpochMilli)
        val startLocalDateTime = LocalDateTime.ofInstant(startInstant, defaultZoneId)
        val endInstant = Instant.ofEpochMilli(endDateEpochMilli)
        val endLocalDateTime = LocalDateTime.ofInstant(endInstant, defaultZoneId)

        Sahha.getSamples(
            sensor = SahhaSensor.valueOf(sensor.uppercase()),
            Pair(startLocalDateTime, endLocalDateTime)
        ) { error, value ->
            if (error != null) {
                if (error.contains("found", ignoreCase = true)) {
                    val data = JSObject()
                    data.put("value", "[]")
                    call.resolve(data)
                } else {
                    call.reject(error)
                }
            } else {
                val gson = GsonBuilder()
                    .registerTypeAdapter(ZonedDateTime::class.java,
                        JsonSerializer<ZonedDateTime> { src, _, _ ->
                            JsonPrimitive(src.toString())
                        }).create()

                val samplesJson = gson.toJson(value)
                val data = JSObject()
                data.put("value", samplesJson)
                call.resolve(data)
            }
        }
    }

    @Deprecated(message = "postSensorData is only supported on iOS", level = DeprecationLevel.WARNING)
    @PluginMethod
    fun postSensorData(call: PluginCall) {
        Log.w(TAG, "postSensorData is only supported on iOS")
        call.resolve()
    }

    @PluginMethod
    fun openAppSettings(call: PluginCall) {
        Sahha.openAppSettings()
        call.resolve()
    }

    fun stringToDrawableResource(context: Context, iconString: String?): Int? {
        return try {
            context.resources.getIdentifier(iconString, "drawable", context.packageName)
        } catch (e: Exception) {
            null
        }
    }
}

File: ios/Sources/SahhaPlugin/SahhaPlugin.swift
import Foundation
import Capacitor
import Sahha

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */

@objc(SahhaPlugin)
public class SahhaPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "SahhaPlugin"
    public let jsName = "Sahha"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAuthenticated", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "authenticate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "authenticateToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deauthenticate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getProfileToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDemographic", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "postDemographic", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSensorStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "enableSensors", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getScores", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBiomarkers", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStats", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSamples", returnType: CAPPluginReturnPromise),
    ]
    
    private func encodeJson<T: Encodable>(_ value: T) throws -> String {
        let jsonData = try JSONEncoder().encode(value)
        if let jsonString = String(data: jsonData, encoding: .utf8) {
            return jsonString
        } else {
            throw EncodingError.invalidValue(value, EncodingError.Context(codingPath: [], debugDescription: "Unable to encode value"))
        }
    }
    
    @objc func configure(_ call: CAPPluginCall) {
        guard let settings = call.getObject("settings") else {
            call.reject("Sahha configure settings parameter is missing")
            return
        }
        
        guard let environment = settings["environment"] as? String, let sahhaEnvironment = SahhaEnvironment(rawValue: environment) else {
            call.reject("Sahha configure settings environment parameter is missing")
            return
        }
        var sahhaSettings = SahhaSettings(environment: sahhaEnvironment)
        sahhaSettings.framework = .capacitor
        
        Sahha.configure(sahhaSettings) {
            call.resolve([
                "success": true
            ])
        }
    }
    
    @objc func isAuthenticated(_ call: CAPPluginCall) {
        call.resolve([
            "success": Sahha.isAuthenticated
        ])
    }
    
    @objc func authenticate(_ call: CAPPluginCall) {
        guard let appId = call.getString("appId") else {
            call.reject("Sahha authenticate appId parameter is missing")
            return
        }
        guard let appSecret = call.getString("appSecret") else {
            call.reject("Sahha authenticate appSecret parameter is missing")
            return
        }
        guard let externalId = call.getString("externalId") else {
            call.reject("Sahha authenticate externalId parameter is missing")
            return
        }
        
        Sahha.authenticate(appId: appId, appSecret: appSecret, externalId: externalId) { error, success in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve([
                    "success": success
                ])
            }
        }
    }
    
    @objc func authenticateToken(_ call: CAPPluginCall) {
        guard let profileToken = call.getString("profileToken") else {
            call.reject("Sahha authenticateToken profileToken parameter is missing")
            return
        }
        guard let refreshToken = call.getString("refreshToken") else {
            call.reject("Sahha authenticateToken refreshToken parameter is missing")
            return
        }
        
        Sahha.authenticate(profileToken: profileToken, refreshToken: refreshToken) { error, success in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve([
                    "success": success
                ])
            }
        }
    }
    
    @objc func deauthenticate(_ call: CAPPluginCall) {
        Sahha.deauthenticate { error, success in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve([
                    "success": success
                ])
            }
        }
    }
    
    @objc func getDemographic(_ call: CAPPluginCall) {
        Sahha.getDemographic { error, value in
            if let error = error {
                call.reject(error)
            } else if let value = value {
                do {
                    let jsonString = try self.encodeJson(value)
                    call.resolve(["demographic": jsonString])
                } catch let encodingError {
                    print(encodingError)
                    call.reject(encodingError.localizedDescription)
                }
            } else {
                call.reject("Sahha demographic is not available")
            }
        }
    }
    
    @objc func postDemographic(_ call: CAPPluginCall) {
        guard let demographic = call.getObject("demographic") else {
            call.reject("Sahha postDemographic demographic parameter is missing")
            return
        }
        
        var sahhaDemographic = SahhaDemographic()
        
        if let ageNumber = demographic["age"] as? NSNumber {
            let age = ageNumber.intValue
            sahhaDemographic.age = age
        }
        
        if let gender = demographic["gender"] as? String {
            sahhaDemographic.gender = gender
        }
        
        if let country = demographic["country"] as? String {
            sahhaDemographic.country = country
        }
        
        if let birthCountry = demographic["birthCountry"] as? String {
            sahhaDemographic.birthCountry = birthCountry
        }
        
        if let ethnicity = demographic["ethnicity"] as? String {
            sahhaDemographic.ethnicity = ethnicity
        }
        
        if let occupation = demographic["occupation"] as? String {
            sahhaDemographic.occupation = occupation
        }
        
        if let industry = demographic["industry"] as? String {
            sahhaDemographic.industry = industry
        }
        
        if let incomeRange = demographic["incomeRange"] as? String {
            sahhaDemographic.incomeRange = incomeRange
        }
        
        if let education = demographic["education"] as? String {
            sahhaDemographic.education = education
        }
        
        if let relationship = demographic["relationship"] as? String {
            sahhaDemographic.relationship = relationship
        }
        
        if let locale = demographic["locale"] as? String {
            sahhaDemographic.locale = locale
        }
        
        if let livingArrangement = demographic["livingArrangement"] as? String {
            sahhaDemographic.livingArrangement = livingArrangement
        }
        
        if let birthDate = demographic["birthDate"] as? String {
            sahhaDemographic.birthDate = birthDate
        }
        
        Sahha.postDemographic(sahhaDemographic) { error, success in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve(["success" : success])
            }
        }
    }
    
    @objc func getSensorStatus(_ call: CAPPluginCall) {
        guard let sensors = call.getArray("sensors") as? [String] else {
            call.reject("Sahha getSensorStatus sensors parameter is missing")
            return
        }
        
        var configSensors: Set<SahhaSensor> = []
        for sensor in sensors {
            if let configSensor = SahhaSensor(rawValue: sensor) {
                configSensors.insert(configSensor)
            }
        }
        
        Sahha.getSensorStatus(configSensors) { error, sensorStatus in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve(["status" : sensorStatus.rawValue])
            }
        }
    }
    
    @objc func enableSensors(_ call: CAPPluginCall) {
        guard let sensors = call.getArray("sensors") as? [String] else {
            call.reject("Sahha enableSensors sensors parameter is missing")
            return
        }
        
        var configSensors: Set<SahhaSensor> = []
        for sensor in sensors {
            if let configSensor = SahhaSensor(rawValue: sensor) {
                configSensors.insert(configSensor)
            }
        }
        
        Sahha.enableSensors(configSensors) { error, sensorStatus in
            if let error = error {
                call.reject(error)
            } else {
                call.resolve(["status" : sensorStatus.rawValue])
            }
        }
    }
    
    @objc func getScores(_ call: CAPPluginCall) {
        guard let types = call.getArray("types", String.self) else {
            call.reject("Sahha getScores types parameter is missing")
            return
        }
        
        guard let startDateEpochMilli = call.getDouble("startDateTime") else {
            call.reject("Sahha getScores startDateTime parameter is missing")
            return
        }
        
        guard let endDateEpochMilli = call.getDouble("endDateTime") else {
            call.reject("Sahha getScores endDateTime parameter is missing")
            return
        }
        
        var sahhaScoreTypes = Set<SahhaScoreType>()
        for type in types {
            if let scoreType = SahhaScoreType(rawValue: type) {
                sahhaScoreTypes.insert(scoreType)
            } else {
                call.reject("Sahha score type parameter is invalid")
                return
            }
        }
        
        let startDate = Date(timeIntervalSince1970: startDateEpochMilli / 1000)
        let endDate = Date(timeIntervalSince1970: endDateEpochMilli / 1000)
        
        Sahha.getScores(types: sahhaScoreTypes, startDateTime: startDate, endDateTime: endDate) { error, scores in
            if let error = error {
                call.reject(error)
            } else if let scores = scores {
                call.resolve(["value": scores])
            } else {
                call.reject("No scores found")
            }
        }
    }
    
    @objc func getBiomarkers(_ call: CAPPluginCall) {
        guard let categories = call.getArray("categories", String.self) else {
            call.reject("Sahha getBiomarkers categories parameter is missing")
            return
        }
        
        guard let types = call.getArray("types", String.self) else {
            call.reject("Sahha getBiomarkers types parameter is missing")
            return
        }
        
        guard let startDateEpochMilli = call.getDouble("startDateTime") else {
            call.reject("Sahha getBiomarkers startDateTime parameter is missing")
            return
        }
        
        guard let endDateEpochMilli = call.getDouble("endDateTime") else {
            call.reject("Sahha getBiomarkers endDateTime parameter is missing")
            return
        }
        
        var sahhaBiomarkerCategories = Set<SahhaBiomarkerCategory>()
        for category in categories {
            if let biomarkerCategory = SahhaBiomarkerCategory(rawValue: category) {
                sahhaBiomarkerCategories.insert(biomarkerCategory)
            } else {
                call.reject("Sahha biomarker category parameter is invalid")
                return
            }
        }
        
        var sahhaBiomarkerTypes = Set<SahhaBiomarkerType>()
        for type in types {
            if let biomarkerType = SahhaBiomarkerType(rawValue: type) {
                sahhaBiomarkerTypes.insert(biomarkerType)
            } else {
                call.reject("Sahha biomarker type parameter is invalid")
                return
            }
        }
        
        let startDate = Date(timeIntervalSince1970: startDateEpochMilli / 1000)
        let endDate = Date(timeIntervalSince1970: endDateEpochMilli / 1000)
        
        Sahha.getBiomarkers(categories: sahhaBiomarkerCategories, types: sahhaBiomarkerTypes, startDateTime: startDate, endDateTime: endDate) { error, biomarkers in
            if let error = error {
                call.reject(error)
            } else if let biomarkers = biomarkers {
                call.resolve(["value": biomarkers])
            } else {
                call.reject("No biomarkers found")
            }
        }
    }
    
    @objc func getStats(_ call: CAPPluginCall) {
        guard let sensor = call.getString("sensor") else {
            call.reject("Sahha getStats sensor parameter is missing")
            return
        }
        
        guard let startDateEpochMilli = call.getDouble("startDateTime") else {
            call.reject("Sahha getStats startDateTime parameter is missing")
            return
        }
        
        guard let endDateEpochMilli = call.getDouble("endDateTime") else {
            call.reject("Sahha getStats endDateTime parameter is missing")
            return
        }
        
        let startDate = Date(timeIntervalSince1970: startDateEpochMilli / 1000)
        let endDate = Date(timeIntervalSince1970: endDateEpochMilli / 1000)
        
        if let sahhaSensor = SahhaSensor(rawValue: sensor) {
            Sahha.getStats(sensor: sahhaSensor, startDateTime: startDate, endDateTime: endDate) { error, stats in
                if let error = error {
                    call.reject(error)
                } else if !stats.isEmpty {
                    do {
                        let jsonEncoder = JSONEncoder()
                        jsonEncoder.outputFormatting = .prettyPrinted
                        let jsonData = try jsonEncoder.encode(stats)
                        if let string = String(data: jsonData, encoding: .utf8) {
                            call.resolve(["value": string])
                        } else {
                            call.reject("Encoding error")
                        }
                    } catch let encodingError {
                        print(encodingError)
                        Sahha.postError(
                            framework: .capacitor,
                            message: encodingError.localizedDescription,
                            path: "SahhaCapacitor", method: "getStats",
                            body: "jsonEncoder")
                        call.reject(encodingError.localizedDescription)
                        return
                    }
                    
                } else {
                    call.reject("No stats found")
                }
            }
        } else {
            call.reject("Invalid sensor: \(sensor)")
        }
    }
    
    @objc func getSamples(_ call: CAPPluginCall) {
        guard let sensor = call.getString("sensor") else {
            call.reject("Sahha getSamples sensor parameter is missing")
            return
        }
        
        guard let startDateEpochMilli = call.getDouble("startDateTime") else {
            call.reject("Sahha getSamples startDate parameter is missing")
            return
        }
        
        guard let endDateEpochMilli = call.getDouble("endDateTime") else {
            call.reject("Sahha getSamples endDate parameter is missing")
            return
        }
        
        let startDate = Date(timeIntervalSince1970: startDateEpochMilli / 1000)
        let endDate = Date(timeIntervalSince1970: endDateEpochMilli / 1000)
        
        if let sahhaSensor = SahhaSensor(rawValue: sensor) {
            Sahha.getSamples(sensor: sahhaSensor, startDateTime: startDate, endDateTime: endDate) { error, stats in
                if let error = error {
                    call.reject(error)
                } else if !stats.isEmpty {
                    do {
                        let jsonEncoder = JSONEncoder()
                        jsonEncoder.outputFormatting = .prettyPrinted
                        let jsonData = try jsonEncoder.encode(stats)
                        if let string = String(data: jsonData, encoding: .utf8) {
                            call.resolve(["value": string])
                        } else {
                            call.reject("Encoding error")
                        }
                    } catch let encodingError {
                        print(encodingError)
                        Sahha.postError(
                            framework: .capacitor,
                            message: encodingError.localizedDescription,
                            path: "SahhaCapacitor", method: "getSamples",
                            body: "jsonEncoder")
                        call.reject(encodingError.localizedDescription)
                        return
                    }
                    
                } else {
                    call.reject("No samples found")
                }
            }
        } else {
            call.reject("Invalid sensor: \(sensor)")
        }
    }
    
    @objc func postSensorData(_ call: CAPPluginCall) {
        Sahha.postSensorData()
    }
    
    @objc func openAppSettings(_ call: CAPPluginCall) {
        Sahha.openAppSettings()
    }
}
